src/btree/bt_curnext.c

```diff
@@ -603,18 +603,26 @@ __wt_btcur_next(WT_CURSOR_BTREE *cbt, bool truncating)
 	 */
 	F_CLR(cbt, WT_CBT_RETRY_PREV);
 	if (F_ISSET(cbt, WT_CBT_RETRY_NEXT)) {
-		WT_RET(__wt_cursor_valid(cbt, &upd, &valid));
+		WT_ERR(__wt_cursor_valid(cbt, &upd, &valid));
+		WT_ASSERT(session, valid == true);
+
+		/* The update that returned prepared conflict is now visible. */
 		F_CLR(cbt, WT_CBT_RETRY_NEXT);
-		if (valid) {
-			/*
-			 * If the update, which returned prepared conflict is
-			 * visible, return the value.
-			 */
-			return (__cursor_kv_return(session, cbt, upd));
-		}
+
+		/*
+		 * The underlying key-return function uses a comparison value
+		 * of 0 to indicate the search function has pre-built the key
+		 * we want to return. That's not the case, don't take that path.
+		 */
+		cbt->compare = 1;
```

Comment:
That was deliberate.

Basically, the **correct** value is 0, because we're pointing to a key which is an exact match for what we're "searching for". The problem is that "searching for" doesn't have a lot of meaning for a scan.

We never called `__cursor_kv_return()` from the cursor scan code before it was added to support prepare-conflict.  The `__cursor_kv_return()` function wasn't ever intended for scans, it was there for updates and other cursor operations that involve searches, and adding it got us into trouble.

The code I'm trying to avoid is here:
```
__key_return(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt)
{               
        WT_CURSOR *cursor;
        WT_ITEM *tmp;
        WT_PAGE *page;
        WT_ROW *rip;
                        
        page = cbt->ref->page;
        cursor = &cbt->iface;   
                                
        if (page->type == WT_PAGE_ROW_LEAF) {
                rip = &page->pg_row[cbt->slot];
                                
                /*      
                 * If the cursor references a WT_INSERT item, take its key.
                 * Else, if we have an exact match, we copied the key in the
                 * search function, take it from there.
                 * If we don't have an exact match, take the key from the
                 * original page.
                 */             
                if (cbt->ins != NULL) {
                        cursor->key.data = WT_INSERT_KEY(cbt->ins);
                        cursor->key.size = WT_INSERT_KEY_SIZE(cbt->ins);
                        return (0); 
                }               
                        
                if (cbt->compare == 0) {         <<<< THIS TEST
```
And, setting `compare = 1/-1` doesn't make any difference for avoiding that test, it just can't be 0.

To be clear, I agree with the change to call `__cursor_kv_return()` in the scan code. The alternative to calling `__cursor_kv_return()` is a lot worse (at least I wasn't able to find anything l liked better when I was messing around with it today). You either end up duplicating a bunch of code, or trying to set the key BEFORE returning a prepare-conflict error on the value, and neither of those led me anywhere I wanted to go.

So, in summary, it doesn't matter to what we set `compare` in the scan code, and setting it to `-1` explicitly seemed misleading, at best.

If you can think of a way to refactor this that would make it more explicit, I'd be a fan. I don't like how we're handing it at the moment, it's seems fragile and dangerous.

But, as I said, when I messed with it today, I didn't find anything I liked better. (The hard part isn't avoiding the test, for example, we could add a cursor field that indicates search has built a key for us. The problem is figuring out when/where to clear that new field, and that gets hard.)

The only approach I didn't hate was changing `__cursor_kv_return()` and all its children (down to `__key_return()`), to take an additional argument to configure if the `cbt->compare == 0` test should be performed, that is, an argument that says "the current cursor position is the result of a search". That wasn't too bad -- I don't honestly remember exactly why I decided against it at the end.

If you like that better, let's talk!